### PR TITLE
CppLink Resource Estimation CPU and min memory options

### PR DESCRIPTION
**Pull request by chiragramani**
This PR makes the allocation of CPU cores and minimum memory for CppLink actions configurable. The motivation behind this change stems from the unique requirements of certain linkers, such as lld and ld_prime, which are multi-threaded and aggressively utilize available CPU cores. In our CI environments, we have encountered issues where multiple parallel link actions compete for resources, leading to linker crashes and build failures.

The crux of the problem lies in achieving a balance between optimizing linker performance and avoiding resource contention. While adjusting the overall concurrency(`--jobs`) could potentially address the issue, it has the unintended consequence of slowing down build times for single-threaded compilation targets. This trade-off is undesirable, as it negatively impacts overall build efficiency.

To address this challenge, this PR introduces two key configurational parameters:

1. **CPU Core Allocation**: Users can now specify the number of CPU cores allocated for linker actions, allowing us to control the degree of parallelism in line with our system's capacity and stability. (also depends on the linker that is being used) **Summary:** Currently, it defaults to 1, but this PR makes it configurable.

2. **Minimum Memory Requirement:** The minimum memory allocated for each linker action is now configurable. This allows us to ensure that link actions have the necessary resources to execute without crashing, while also minimizing the potential for resource bottlenecks. **Summary**: Earlier it was always 15 Mb for macOS, which certainly would be different for different linkers. This PR makes it configurable.

**Note:**
1. This only affects local execution.
2. From my understanding, execution_requirements are only applicable to RBE and not for local execution, hence I didn't explore that direction, but can surely be extended.
3. This PR only makes the existing values configurable and doesn't change any default assumption/logic.

**Repro of the working post applying this PR**
Requirements: macOS, Xcode 15.
1. Download [ExecMemRepro-gen.zip](https://github.com/bazelbuild/bazel/files/12972707/ExecMemRepro-gen.zip)
2. The below was executed on a 10 core machine.
a.) LinkAction taking all cores.

Here we see, that a link action would be taking all the cores =10, and from the trace, note only one link action happened at a time and Bazel was in an `Acquiring resources mode` queuing up other parallel link actions.
```
./bazelc build //:UnitTests //:UnitTests1 //:UnitTests2 //:UnitTests3 //:UnitTests4 //:UnitTests5 --experimental_cpp_link_resources_estimation_cpu=10 --profile=tencores.trace.gz
```
Screenshot 2023-10-17 at 5 36 59 PM

[tencores.trace.gz](https://github.com/bazelbuild/bazel/files/12972732/tencores.trace.gz)

b.) LinkAction taking only one core.
```
./bazelc build //:UnitTests //:UnitTests1 //:UnitTests2 //:UnitTests3 //:UnitTests4 //:UnitTests5 --experimental_cpp_link_resources_estimation_cpu=1 --profile=singlecore.trace.gz
```
Here we see, that all link actions are in parallel.

[single-core.trace.gz](https://github.com/bazelbuild/bazel/files/12972728/single-core.trace.gz)
Screenshot 2023-10-17 at 5 35 19 PM

I'm open to the possibility of more efficient ways to make this configuration flexible, please let me know; I would be more than happy to explore those options further.

**TLDR:** CppLinkActions currently have a default non-configurable estimation where ResourceSet always evaluates to single core CPU and some default platform-specific minimum memory thresholds. This PR makes few of these attributes configurable.

### CODE REVIEW

**Comment by meisterT** [comment]
How much will this actually help with wall time on the critical path?

I wonder how this might interact with C++ Starlarkification? cc @buildbreaker2021

**Comment by tjgq** [comment]
Do we really need to introduce a flag for this? Can this be done with `--modify_execution_requirements` instead? If not, can/should we add a general capability instead of a CppLink-specific one?

**Comment by comius** [comment]
I’m a little late to the party. As I’m currently working on starlarkifying CppLinkAction, this would create additional work. Also we have so many flag / knobs and buttons already, that a preferred solution would work without them. That is, well chosen defaults that work in most cases.

CppLinkAction is already dumping information that should make it possible to fit such data.
